GG/src/Font.cpp

```diff
@@ -1253,9 +1377,238 @@ Pt Font::DetermineLinesImpl(const std::string& text,
+                       const X box_width,
+                       const std::vector<Font::LineData>& line_data) {
+        std::cout << "Font::DetermineLines(text=\"" << text << "\" (@ "
+                  << static_cast<const void*>(&*text.begin()) << ") format="
+                  << format << " box_width=" << box_width << ")" << std::endl;
+
+        std::cout << "Line breakdown:\n";
+        for (std::size_t i = 0; i < line_data.size(); ++i) {
+            std::cout << "Line " << i << ":\n    extents=";
+            for (std::size_t j = 0; j < line_data[i].char_data.size(); ++j) {
+                std::cout << line_data[i].char_data[j].extent << " ";
+            }
+            std::cout << "\n    string indices=";
+            for (std::size_t j = 0; j < line_data[i].char_data.size(); ++j) {
+                std::cout << line_data[i].char_data[j].string_index << " ";
+            }
+            std::cout << "\n    code point indices=";
+            for (std::size_t j = 0; j < line_data[i].char_data.size(); ++j) {
+                std::cout << line_data[i].char_data[j].code_point_index << " ";
+            }
+            std::cout << "\n    chars on line: \"";
+            for (std::size_t j = 0; j < line_data[i].char_data.size(); ++j) {
+                std::cout << text[Value(line_data[i].char_data[j].string_index)];
+            }
+            std::cout << "\"" << std::endl;
+            for (std::size_t j = 0; j < line_data[i].char_data.size(); ++j) {
+                for (std::size_t k = 0; k < line_data[i].char_data[j].tags.size(); ++k) {
+                    if (boost::shared_ptr<Font::FormattingTag> tag_elem = line_data[i].char_data[j].tags[k]) {
+                        std::cout << "FormattingTag @" << j << "\n    text=\"" << tag_elem->text << "\"\n    widths=";
+                        for (std::size_t j = 0; j < tag_elem->widths.size(); ++j) {
+                            std::cout << tag_elem->widths[j] << " ";
+                        }
+                        std::cout << "\n    whitespace=" << tag_elem->whitespace
+                                  << "\n    newline=" << tag_elem->newline << "\n    params=\n";
+                        for (std::size_t l = 0; l < tag_elem->params.size(); ++l) {
+                            std::cout << "        \"" << tag_elem->params[l] << "\"\n";
+                        }
+                        std::cout << "    tag_name=\"" << tag_elem->tag_name << "\"\n    close_tag="
+                                  << tag_elem->close_tag << std::endl;
+                    }
+                }
+            }
+            std::cout << "    justification=" << line_data[i].justification << "\n" << std::endl;
+        }
+    }
+}
+
+std::vector<boost::shared_ptr<Font::TextElement> > Font::ComputeTextElements(
+    const std::string& text,
+    const Flags<TextFormat>& format) const
+{
+    std::vector<boost::shared_ptr<TextElement> > text_elements;
```

Comment:
can you add a bunch more comments throughout all this code to explain what it's doing?